pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java

```diff
@@ -45,19 +62,50 @@ public static MetadataStoreExtended createExtended(String metadataURL, MetadataS
+        }
+        return new ZKMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
+    }
+
+    static void loadProviders() {
+        String factoryClasses = System.getProperty(METADATASTORE_PROVIDERS_PROPERTY);
+        providers = new HashMap<>();
+        providers.put(MEMORY_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(MEMORY_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> new LocalMemoryMetadataStore(url, config)));
+        providers.put(ROCKSDB_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ROCKSDB_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> RocksdbMetadataStore.get(url, config)));
+        providers.put(ETCD_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ETCD_SCHEME_IDENTIFIER,
+                EtcdMetadataStore::new));
+        providers.put(ZK_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ZK_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) ->
+                        new ZKMetadataStore(url.substring(ZK_SCHEME_IDENTIFIER.length()), config, enableWatchers)));
+
+        if (factoryClasses == null) {
+            return;
+        }
+        String[] classNames = factoryClasses.split(",");
+        for (String className : classNames) {
+            try {
+                Class<? extends MetadataStoreProvider> clazz =
+                        (Class<? extends MetadataStoreProvider>) Class.forName(className);
+                MetadataStoreProvider provider = clazz.getConstructor().newInstance();
+                String scheme = provider.urlScheme();
+                providers.put(scheme, provider);
+            } catch (Exception e) {
+                log.warn("Failed to load metadata store provider class for name '{}'", className, e);
+            }
+        }
+    }
 
-        if (metadataURL.startsWith(LocalMemoryMetadataStore.MEMORY_SCHEME_IDENTIFIER)) {
-            return new LocalMemoryMetadataStore(metadataURL, metadataStoreConfig);
-        } else if (metadataURL.startsWith(RocksdbMetadataStore.ROCKSDB_SCHEME_IDENTIFIER)) {
-            return RocksdbMetadataStore.get(metadataURL, metadataStoreConfig);
-        } else if (metadataURL.startsWith(EtcdMetadataStore.ETCD_SCHEME_IDENTIFIER)) {
-            return new EtcdMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
-        } else if (metadataURL.startsWith(ZKMetadataStore.ZK_SCHEME_IDENTIFIER)) {
-            return new ZKMetadataStore(metadataURL.substring(ZKMetadataStore.ZK_SCHEME_IDENTIFIER.length()),
-                    metadataStoreConfig, enableSessionWatcher);
-        } else {
-            return new ZKMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
+    private static MetadataStoreProvider findProvider(String metadataURL) {
+        for (Map.Entry<String, MetadataStoreProvider> entry : providers.entrySet()) {
+            if (metadataURL.startsWith(entry.getKey())) {
```

Comment:
Something doesn't add up here: When you register the known providers, you place them in the map with the colons:

```
providers.put(ETCD_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ETCD_SCHEME_IDENTIFIER,
```

```
static final String ETCD_SCHEME_IDENTIFIER = "etcd:";
```

In the interface someone needs to implemented it has:

```
public interface MetadataStoreProvider {


    /**
     * Scheme of the urls that MetadataSores provided by this object can handle.
     */
    String urlScheme();
```

Normally a scheme is just a name without colons, so I believe this will be confusing.


I suggest:

1. Use the scheme names in the map.
2. Interface method `urlScheme()` will return just scheme names without colons
3. Search from the beginning of the string to the first occurrence of ":"; get that string without colons, which is your scheme.
4. If the scheme doesn't exist, assume the scheme is "zk" and retrieve based on it.